diff --git a/chirpstack-concentratord-sx1301/src/config/mod.rs b/chirpstack-concentratord-sx1301/src/config/mod.rs
index c747a07..904841e 100644
--- a/chirpstack-concentratord-sx1301/src/config/mod.rs
+++ b/chirpstack-concentratord-sx1301/src/config/mod.rs
@@ -209,7 +209,8 @@ pub fn get(filenames: Vec<String>) -> Configuration {
         "rak_2247" => vendor::rak::rak2247::new(&config).unwrap(),
         "risinghf_rhf0m301" => vendor::risinghf::rhf0m301::new(&config).unwrap(),
         "sandbox_lorago_port" => vendor::sandbox::lorago_port::new(&config).unwrap(),
-        "wifx_lorix_one" => vendor::wifx::lorix_one::new(&config).unwrap(),
+        "wifx_lorix_one_863_870" => vendor::wifx::lorix_one_863_870::new(&config).unwrap(),
+        "wifx_lorix_one_902_928" => vendor::wifx::lorix_one_902_928::new(&config).unwrap(),
         _ => panic!("unexpected gateway model: {}", config.gateway.model),
     };
 
diff --git a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one.rs b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one.rs
deleted file mode 100644
index ebf6f46..0000000
--- a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one.rs
+++ /dev/null
@@ -1,296 +0,0 @@
-use anyhow::Result;
-use libloragw_sx1301::hal;
-
-use super::super::super::super::config::{self, Region};
-use super::super::{Configuration, Gps};
-
-// source:
-// https://github.com/Wifx/meta-wifx/blob/krogoth/recipes-connectivity/packet-forwarder/files/configs/global_conf_EU868_2dBi_indoor.json
-// https://github.com/Wifx/meta-wifx/blob/krogoth/recipes-connectivity/packet-forwarder/files/configs/global_conf_EU868_4dBi_outdoor.json
-pub fn new(conf: &config::Configuration) -> Result<Configuration> {
-    let region = conf.gateway.region.unwrap_or(Region::EU868);
-
-    let radio_min_max_tx_freq = match region {
-        Region::EU868 => vec![(863000000, 870000000), (863000000, 870000000)],
-        _ => return Err(anyhow!("Region is not supported: {}", region)),
-    };
-
-    Ok(Configuration {
-        radio_min_max_tx_freq,
-        radio_count: 2,
-        clock_source: 1,
-        radio_rssi_offset: vec![-164.0, -164.0],
-        radio_tx_enabled: vec![true, false],
-        radio_type: vec![hal::RadioType::SX1257, hal::RadioType::SX1257],
-        radio_tx_notch_freq: vec![129000, 0],
-        lora_multi_sf_bandwidth: 125000,
-        tx_gain_table: if conf.gateway.antenna_gain == 2 {
-            vec![
-                // 0
-                hal::TxGainConfig {
-                    pa_gain: 0,
-                    mix_gain: 10,
-                    rf_power: -6,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 1
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 9,
-                    rf_power: -3,
-                    dig_gain: 3,
-                    dac_gain: 3,
-                },
-                // 2
-                hal::TxGainConfig {
-                    pa_gain: 0,
-                    mix_gain: 15,
-                    rf_power: 0,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 3
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 9,
-                    rf_power: 3,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 4
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 8,
-                    rf_power: 6,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 5
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 10,
-                    rf_power: 10,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 6
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 9,
-                    rf_power: 11,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 7
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 11,
-                    rf_power: 12,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 8
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 11,
-                    rf_power: 13,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 9
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 12,
-                    rf_power: 14,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 10
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 12,
-                    rf_power: 16,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 11
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 12,
-                    rf_power: 20,
-                    dig_gain: 3,
-                    dac_gain: 3,
-                },
-                // 12
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 11,
-                    rf_power: 23,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 13
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 12,
-                    rf_power: 25,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 14
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 12,
-                    rf_power: 26,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 15
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 13,
-                    rf_power: 27,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-            ]
-        } else if conf.gateway.antenna_gain == 4 {
-            vec![
-                // 0
-                hal::TxGainConfig {
-                    pa_gain: 0,
-                    mix_gain: 9,
-                    rf_power: -6,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 1
-                hal::TxGainConfig {
-                    pa_gain: 0,
-                    mix_gain: 10,
-                    rf_power: -3,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 2
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 8,
-                    rf_power: 0,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 3
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 11,
-                    rf_power: 3,
-                    dig_gain: 3,
-                    dac_gain: 3,
-                },
-                // 4
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 13,
-                    rf_power: 6,
-                    dig_gain: 3,
-                    dac_gain: 3,
-                },
-                // 5
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 13,
-                    rf_power: 10,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 6
-                hal::TxGainConfig {
-                    pa_gain: 1,
-                    mix_gain: 15,
-                    rf_power: 11,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 7
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 10,
-                    rf_power: 12,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 8
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 10,
-                    rf_power: 13,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 9
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 11,
-                    rf_power: 14,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 10
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 11,
-                    rf_power: 16,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 11
-                hal::TxGainConfig {
-                    pa_gain: 2,
-                    mix_gain: 15,
-                    rf_power: 20,
-                    dig_gain: 2,
-                    dac_gain: 3,
-                },
-                // 12
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 10,
-                    rf_power: 23,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 13
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 11,
-                    rf_power: 25,
-                    dig_gain: 1,
-                    dac_gain: 3,
-                },
-                // 14
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 11,
-                    rf_power: 26,
-                    dig_gain: 0,
-                    dac_gain: 3,
-                },
-                // 15
-                hal::TxGainConfig {
-                    pa_gain: 3,
-                    mix_gain: 15,
-                    rf_power: 27,
-                    dig_gain: 3,
-                    dac_gain: 3,
-                },
-            ]
-        } else {
-            panic!("Invalid antenna_gain: {}", conf.gateway.antenna_gain);
-        },
-        gps: Gps::None,
-        spidev_path: "/dev/spidev0.0".to_string(),
-        reset_pin: conf.gateway.get_sx1301_reset_pin("/dev/gpiochip0", 1),
-    })
-}
diff --git a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_863_870.rs b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_863_870.rs
new file mode 100644
index 0000000..42f3fcf
--- /dev/null
+++ b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_863_870.rs
@@ -0,0 +1,90 @@
+use anyhow::Result;
+use libloragw_sx1302::hal;
+
+use super::super::super::super::config;
+use super::super::{ComType, Configuration, Gps, RadioConfig};
+
+pub fn new(conf: &config::Configuration) -> Result<Configuration> {
+    Ok(Configuration {
+        radio_count: 2,
+        clock_source: 1,
+        radio_rssi_offset: vec![-166.0, -166.0],
+        radio_tx_enabled: vec![true, false],
+        radio_type: vec![hal::RadioType::SX1257, hal::RadioType::SX1257],
+        radio_min_max_tx_freq: vec![(863000000, 870000000), (863000000, 870000000)],
+        radio_tx_notch_freq: vec![0, 0],
+        lora_multi_sf_bandwidth: 125000,
+        tx_gain_table: get_tx_gain_table(conf.gateway.antenna_gain),
+        gps: None,
+        spidev_path: "/dev/spidev0.0".to_owned(),
+        reset_pin: conf.gateway.get_sx1301_reset_pin("/dev/gpiochip0", 1)
+    })
+}
+
+// Power table with an antenna gain of 0dBi
+lazy_static! {
+    static ref POWER_TABLE_BASE: [hal::TxGainConfig; 21] = [
+        gain_param(7, 1, 15, 2),
+        gain_param(8, 2, 9, 1),
+        gain_param(9, 2, 12, 3),
+        gain_param(10, 3, 8, 3),
+        gain_param(11, 2, 10, 0),
+        gain_param(12, 3, 9, 3),
+        gain_param(13, 2, 11, 0),
+        gain_param(14, 2, 13, 2),
+        gain_param(15, 3, 8, 1),
+        gain_param(16, 2, 13, 0),
+        gain_param(17, 2, 14, 0),
+        gain_param(18, 3, 10, 2),
+        gain_param(19, 3, 9, 0),
+        gain_param(20, 3, 10, 1),
+        gain_param(21, 3, 13, 3),
+        gain_param(22, 3, 12, 2),
+        gain_param(23, 3, 11, 0),
+        gain_param(24, 3, 13, 2),
+        gain_param(25, 3, 14, 2),
+        gain_param(26, 3, 14, 0),
+        gain_param(27, 3, 15, 0),
+    ];
+}
+
+fn gain_param(rf_power: i8, pa: u8, mix: u8, dig: u8) -> hal::TxGainConfig {
+    hal::TxGainConfig {
+        dig_gain: dig,
+        pa_gain: pa,
+        dac_gain: 3,
+        mix_gain: mix,
+        rf_power: rf_power,
+    }
+}
+
+fn get_tx_gain_table(antenna_gain: i8) -> Vec<hal::TxGainConfig> {
+    // filter the power values
+    let entries: Vec<hal::TxGainConfig> = POWER_TABLE_BASE
+        .iter()
+        .map(|entry| entry.clone())
+        .filter(|entry| {
+            let radiated_power = entry.rf_power + antenna_gain;
+            if radiated_power > 27 {
+                debug!(
+                    "Power table entry including antenna gain is too high, radiated power: {}, discarded",
+                    radiated_power
+                );
+                    return false;
+                }
+            return true;
+        })
+        .collect();
+
+    let count = entries.len();
+    if count == 0 {
+        panic!("No entries available for power table");
+    }
+    let mut start_index = 0;
+    if count > 16 {
+        start_index = count - 16;
+    }
+
+    // keep only max 16 higher entries
+    entries.as_slice()[start_index..].to_vec()
+}
diff --git a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_902_928.rs b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_902_928.rs
new file mode 100644
index 0000000..82972c6
--- /dev/null
+++ b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/lorix_one_902_928.rs
@@ -0,0 +1,90 @@
+use anyhow::Result;
+use libloragw_sx1302::hal;
+
+use super::super::super::super::config;
+use super::super::{ComType, Configuration, Gps, RadioConfig};
+
+pub fn new(conf: &config::Configuration) -> Result<Configuration> {
+    Ok(Configuration {
+        radio_count: 2,
+        clock_source: 1,
+        radio_rssi_offset: vec![-166.0, -166.0],
+        radio_tx_enabled: vec![true, false],
+        radio_type: vec![hal::RadioType::SX1257, hal::RadioType::SX1257],
+        radio_min_max_tx_freq: vec![(902000000, 928000000), (902000000, 928000000)],
+        radio_tx_notch_freq: vec![0, 0],
+        lora_multi_sf_bandwidth: 125000,
+        tx_gain_table: get_tx_gain_table(conf.gateway.antenna_gain),
+        gps: None,
+        spidev_path: "/dev/spidev0.0".to_owned(),
+        reset_pin: conf.gateway.get_sx1301_reset_pin("/dev/gpiochip0", 1)
+    })
+}
+
+// Power table with an antenna gain of 0dBi
+lazy_static! {
+    static ref POWER_TABLE_BASE: [hal::TxGainConfig; 21] = [
+        gain_param(7, 1, 12, 2),
+        gain_param(8, 1, 12, 1),
+        gain_param(9, 1, 12, 0),
+        gain_param(10, 1, 13, 0),
+        gain_param(11, 1, 15, 1),
+        gain_param(12, 2, 9, 0),
+        gain_param(13, 3, 8, 3),
+        gain_param(14, 2, 10, 0),
+        gain_param(15, 3, 9, 3),
+        gain_param(16, 2, 12, 1),
+        gain_param(17, 3, 10, 3),
+        gain_param(18, 2, 14, 2),
+        gain_param(19, 2, 15, 2),
+        gain_param(20, 3, 9, 1),
+        gain_param(21, 3, 12, 3),
+        gain_param(22, 3, 10, 1),
+        gain_param(23, 3, 11, 2),
+        gain_param(24, 3, 11, 1),
+        gain_param(25, 3, 14, 3),
+        gain_param(26, 3, 13, 2),
+        gain_param(27, 3, 13, 1),
+    ];
+}
+
+fn gain_param(rf_power: i8, pa: u8, mix: u8, dig: u8) -> hal::TxGainConfig {
+    hal::TxGainConfig {
+        dig_gain: dig,
+        pa_gain: pa,
+        dac_gain: 3,
+        mix_gain: mix,
+        rf_power: rf_power,
+    }
+}
+
+fn get_tx_gain_table(antenna_gain: i8) -> Vec<hal::TxGainConfig> {
+    // filter the power values
+    let entries: Vec<hal::TxGainConfig> = POWER_TABLE_BASE
+        .iter()
+        .map(|entry| entry.clone())
+        .filter(|entry| {
+            let radiated_power = entry.rf_power + antenna_gain;
+            if radiated_power > 27 {
+                debug!(
+                    "Power table entry including antenna gain is too high, radiated power: {}, discarded",
+                    radiated_power
+                );
+                    return false;
+                }
+            return true;
+        })
+        .collect();
+
+    let count = entries.len();
+    if count == 0 {
+        panic!("No entries available for power table");
+    }
+    let mut start_index = 0;
+    if count > 16 {
+        start_index = count - 16;
+    }
+
+    // keep only max 16 higher entries
+    entries.as_slice()[start_index..].to_vec()
+}
diff --git a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/mod.rs b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/mod.rs
index ac0dab1..4fdecef 100644
--- a/chirpstack-concentratord-sx1301/src/config/vendor/wifx/mod.rs
+++ b/chirpstack-concentratord-sx1301/src/config/vendor/wifx/mod.rs
@@ -1 +1,2 @@
-pub mod lorix_one;
+pub mod lorix_one_863_870;
+pub mod lorix_one_902_928;
