From e6c7049dbdebaf7cc097e83bca931885b4f07c09 Mon Sep 17 00:00:00 2001
From: Yannick Lanz <yannick.lanz@wifx.net>
Date: Thu, 22 Jul 2021 23:23:04 +0200
Subject: [PATCH 5/5] Add mender support for the L1 machine

---
 include/configs/l1.h | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/include/configs/l1.h b/include/configs/l1.h
index af961cb71f..afe6f2ecb1 100644
--- a/include/configs/l1.h
+++ b/include/configs/l1.h
@@ -50,6 +50,14 @@
 #define CONFIG_ATMEL_NAND_HW_PMECC
 #endif
 
+#ifndef CONFIG_EXTRA_ENV_SETTINGS
+#define CONFIG_EXTRA_ENV_SETTINGS \
+	"mtdids=" CONFIG_MTDIDS_DEFAULT "\0" 		\
+	"mtdparts=" CONFIG_MTDPARTS_DEFAULT "\0" 	\
+	"kernel_addr_r=0x22000000\0" 				\
+	"fdt_addr_r=0x21000000\0"
+#endif
+
 /* SPL */
 #define CONFIG_SPL_TEXT_BASE		0x200000
 #define CONFIG_SPL_MAX_SIZE			0x18000
@@ -89,4 +97,8 @@
 #define CONFIG_AT91SAM9_WATCHDOG
 #define CONFIG_HW_WATCHDOG
 
+/* Mender integration */
+#define CONFIG_BOOTCOUNT_LIMIT
+#define CONFIG_BOOTCOUNT_ENV
+
 #endif /* __CONFIG_H */
-- 
2.25.1

