Index: support/modules/rootfs-image
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/support/modules/rootfs-image b/support/modules/rootfs-image
--- a/support/modules/rootfs-image	(revision 121dbbd87c65344679c287b3f1e249a1372f3c4f)
+++ b/support/modules/rootfs-image	(date 1730997545478)
@@ -11,11 +11,14 @@
     JQ_AVAILABLE=0
 fi
 
-if which mender-flash >/dev/null 2>&1; then
-    MENDER_FLASH_AVAILABLE=1
-else
-    MENDER_FLASH_AVAILABLE=0
-fi
+#if which mender-flash >/dev/null 2>&1; then
+#    MENDER_FLASH_AVAILABLE=1
+#else
+#    MENDER_FLASH_AVAILABLE=0
+#fi
+
+# Disable mender-flash as it will fail if run before ubiupdatevol
+MENDER_FLASH_AVAILABLE=0
 
 if command -v grub-mender-grubenv-print > /dev/null; then
     PRINTENV=grub-mender-grubenv-print
