Index: support/CMakeLists.txt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/support/CMakeLists.txt b/support/CMakeLists.txt
--- a/support/CMakeLists.txt	(revision 121dbbd87c65344679c287b3f1e249a1372f3c4f)
+++ b/support/CMakeLists.txt	(date 1730996978898)
@@ -23,12 +23,6 @@
   modules/rootfs-image
 )
 set(MODULES
-  modules/deb
-  modules/docker
-  modules/directory
-  modules/single-file
-  modules/rpm
-  modules/script
   ${ROOTFS_IMAGE}
 )
 set(MODULES_ARTIFACT_GENERATORS
