diff --git a/Cargo.toml b/Cargo.toml
index b772bd7..654b0c0 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,5 +1,3 @@
-cargo-features = ["strip"]
-
 [package]
 name = "gateway-rs"
 description = "Helium Gateway for LoRa packet forwarders"
@@ -56,7 +54,6 @@ codegen-units = 1
 panic = "abort"
 #debug = true
 #strip = "debuginfo"
-strip = "symbols"
 
 [patch.crates-io]
 elliptic-curve = { git = "https://github.com/helium/traits.git", branch = "rg/compact" }
