From b2a4a3cef3fa9b497094c1efa2014f27467a9f72 Mon Sep 17 00:00:00 2001
From: Yannick Lanz <yannick.lanz@wifx.net>
Date: Mon, 31 Mar 2025 11:17:49 +0200
Subject: [PATCH 06/11] mc610: add CID0 PDP context minimal support

---
 .../mm-broadband-modem-fibocom-mc610.c        | 746 +++++++++++++++++-
 1 file changed, 737 insertions(+), 9 deletions(-)

diff --git a/src/plugins/fibocom_mc610/mm-broadband-modem-fibocom-mc610.c b/src/plugins/fibocom_mc610/mm-broadband-modem-fibocom-mc610.c
index cf047c3f..56e9d842 100644
--- a/src/plugins/fibocom_mc610/mm-broadband-modem-fibocom-mc610.c
+++ b/src/plugins/fibocom_mc610/mm-broadband-modem-fibocom-mc610.c
@@ -26,18 +26,48 @@
 #include "mm-base-modem-at.h"
 #include "mm-iface-modem.h"
 #include "mm-iface-modem-firmware.h"
+#include "mm-iface-modem-3gpp.h"
+#include "mm-iface-modem-3gpp-profile-manager.h"
 #include "mm-log.h"
+#include "mm-log-helpers.h"
 #include "mm-shared-fibocom.h"
 
-static void iface_modem_init          (MMIfaceModemInterface         *iface);
-static void iface_modem_firmware_init (MMIfaceModemFirmwareInterface *iface);
-static void shared_fibocom_init       (MMSharedFibocomInterface      *iface);
-
-static MMIfaceModemFirmwareInterface *iface_modem_firmware_parent;
+/* Codes ANSI pour les couleurs */
+#define ANSI_COLOR_RED     "\x1b[31m"
+#define ANSI_COLOR_GREEN   "\x1b[32m"
+#define ANSI_COLOR_YELLOW  "\x1b[33m"
+#define ANSI_COLOR_BLUE    "\x1b[34m"
+#define ANSI_COLOR_MAGENTA "\x1b[35m"
+#define ANSI_COLOR_CYAN    "\x1b[36m"
+#define ANSI_COLOR_RESET   "\x1b[0m"
+
+/* Fonction pour afficher un message de debug en rouge */
+#define mm_obj_dbg_red(self, fmt, ...)                                   \
+    mm_obj_dbg (self, ANSI_COLOR_RED fmt ANSI_COLOR_RESET, ##__VA_ARGS__)
+
+/* Fonction pour afficher un message de debug en jaune */
+#define mm_obj_dbg_yellow(self, fmt, ...)                                \
+    mm_obj_dbg (self, ANSI_COLOR_YELLOW fmt ANSI_COLOR_RESET, ##__VA_ARGS__)
+
+/* Fonction pour afficher un message de debug en vert */
+#define mm_obj_dbg_green(self, fmt, ...)                                 \
+    mm_obj_dbg (self, ANSI_COLOR_GREEN fmt ANSI_COLOR_RESET, ##__VA_ARGS__)
+
+static void iface_modem_init                      (MMIfaceModemInterface                   *iface);
+static void iface_modem_firmware_init             (MMIfaceModemFirmwareInterface           *iface);
+static void iface_modem_3gpp_init                 (MMIfaceModem3gppInterface               *iface);
+static void iface_modem_3gpp_profile_manager_init (MMIfaceModem3gppProfileManagerInterface *iface);
+static void shared_fibocom_init                   (MMSharedFibocomInterface                *iface);
+
+static MMIfaceModemFirmwareInterface           *iface_modem_firmware_parent;
+static MMIfaceModem3gppInterface               *iface_modem_3gpp_parent;
+static MMIfaceModem3gppProfileManagerInterface *iface_modem_3gpp_profile_manager_parent;
 
 G_DEFINE_TYPE_EXTENDED (MMBroadbandModemFibocomMc610, mm_broadband_modem_fibocom_mc610, MM_TYPE_BROADBAND_MODEM, 0,
                         G_IMPLEMENT_INTERFACE (MM_TYPE_IFACE_MODEM, iface_modem_init)
                         G_IMPLEMENT_INTERFACE (MM_TYPE_SHARED_FIBOCOM,  shared_fibocom_init)
+                        G_IMPLEMENT_INTERFACE (MM_TYPE_IFACE_MODEM_3GPP, iface_modem_3gpp_init)
+                        G_IMPLEMENT_INTERFACE (MM_TYPE_IFACE_MODEM_3GPP_PROFILE_MANAGER, iface_modem_3gpp_profile_manager_init)
                         G_IMPLEMENT_INTERFACE (MM_TYPE_IFACE_MODEM_FIRMWARE, iface_modem_firmware_init))
 
 typedef enum {
@@ -61,6 +91,9 @@ struct _MMBroadbandModemFibocomMc610Private {
 
     /* Initial EPS bearer context number */
     gint            initial_eps_bearer_cid;
+
+    MM3gppProfile  *profile0;
+    gboolean        profile0_stored;
 };
 
 /*****************************************************************************/
@@ -318,7 +351,7 @@ set_current_modes_finish (MMIfaceModem  *self,
 }
 
 static void
-allowed_access_technology_update_ready (MMBroadbandModemFibocom *self,
+allowed_access_technology_update_ready (MMBroadbandModemFibocomMc610 *self,
                                         GAsyncResult            *res,
                                         GTask                   *task)
 {
@@ -570,6 +603,670 @@ modem_cleanup_sim_hot_swap (MMIfaceModem *self)
     mm_broadband_modem_sim_hot_swap_ports_context_reset (MM_BROADBAND_MODEM (self));
 }
 
+/*****************************************************************************/
+/* Initial EPS bearer settings loading -> set configuration */
+
+typedef enum {
+    BEARER_FIBOCOM_AUTH_UNKNOWN   = -1,
+    BEARER_FIBOCOM_AUTH_NONE      =  0,
+    BEARER_FIBOCOM_AUTH_PAP       =  1,
+    BEARER_FIBOCOM_AUTH_CHAP      =  2,
+} BearerFibocomAuthType;
+
+static MMBearerAllowedAuth
+mm_auth_type_from_fibocom_auth_type (guint fibocom_auth)
+{
+    switch (fibocom_auth) {
+    case BEARER_FIBOCOM_AUTH_NONE:
+        return MM_BEARER_ALLOWED_AUTH_NONE;
+    case BEARER_FIBOCOM_AUTH_PAP:
+        return MM_BEARER_ALLOWED_AUTH_PAP;
+    case BEARER_FIBOCOM_AUTH_CHAP:
+        return MM_BEARER_ALLOWED_AUTH_CHAP;
+    default:
+        return MM_BEARER_ALLOWED_AUTH_UNKNOWN;
+    }
+}
+
+static gboolean
+mm_fibocom_parse_mgauth_response (const gchar          *response,
+    guint                 cid,
+    MMBearerAllowedAuth  *out_auth,
+    GError              **error)
+{
+    g_autoptr(GRegex)     r = NULL;
+    g_autoptr(GMatchInfo) match_info = NULL;
+
+    r = g_regex_new ("\\+MGAUTH:\\s*(\\d+),(\\d+)", 0, 0, NULL);
+    g_assert (r != NULL);
+
+    g_regex_match_full (r, response, strlen (response), 0, 0, &match_info, NULL);
+    while (g_match_info_matches (match_info)) {
+        guint mgauth_cid = 0;
+
+        if (mm_get_uint_from_match_info (match_info, 1, &mgauth_cid) &&
+            (mgauth_cid == cid)) {
+            guint fibocom_auth_type = 0;
+
+            mm_get_uint_from_match_info (match_info, 2, &fibocom_auth_type);
+            *out_auth = mm_auth_type_from_fibocom_auth_type (fibocom_auth_type);
+            return TRUE;
+        }
+        g_match_info_next (match_info, NULL);
+    }
+
+    g_set_error (error, MM_CORE_ERROR, MM_CORE_ERROR_NOT_FOUND,
+                    "Auth settings for context %u not found", cid);
+    return FALSE;
+}
+
+static MMBearerProperties *
+modem_3gpp_load_initial_eps_bearer_settings_finish (MMIfaceModem3gpp  *self,
+                                                    GAsyncResult      *res,
+                                                    GError           **error)
+{
+    return g_task_propagate_pointer (G_TASK (res), error);
+}
+
+static void
+load_initial_eps_bearer_settings_auth_ready (MMBaseModem  *self,
+                                             GAsyncResult *res,
+                                             GTask        *task)
+{
+    const gchar         *response;
+    g_autoptr(GError)    error = NULL;
+    MMBearerAllowedAuth  auth = MM_BEARER_ALLOWED_AUTH_UNKNOWN;
+    g_autofree gchar    *username = NULL;
+    gint                 cid;
+    MMBearerProperties  *properties;
+
+    mm_obj_dbg_red (self, "load_initial_eps_bearer_settings_auth_ready");
+
+    properties = g_task_get_task_data (task);
+
+    cid = mm_broadband_modem_get_initial_eps_bearer_cid (MM_BROADBAND_MODEM (self));
+    g_assert (cid >= 0);
+
+    response = mm_base_modem_at_command_finish (self, res, &error);
+    if (!response)
+        mm_obj_dbg_red (self, "couldn't load auth settings: %s", error->message);
+    else if (!mm_fibocom_parse_mgauth_response (response, cid, &auth, &error))
+        mm_obj_dbg_red (self, "couldn't parse auth settings for cid %d: %s", cid, error->message);
+    else {
+        mm_bearer_properties_set_allowed_auth (properties, auth);
+    }
+
+    g_task_return_pointer (task, g_object_ref (properties), g_object_unref);
+    g_object_unref (task);
+}
+
+static void
+load_initial_eps_bearer_settings_parent_ready (MMIfaceModem3gpp *_self,
+                                               GAsyncResult     *res,
+                                               GTask            *task)
+{
+    MMBroadbandModemFibocomMc610 *self = MM_BROADBAND_MODEM_FIBOCOM_MC610 (_self);
+    MMBearerProperties           *properties;
+    GError                       *error = NULL;
+
+    mm_obj_dbg_red (self, "load_initial_eps_bearer_settings_parent_ready");
+
+    properties = iface_modem_3gpp_parent->load_initial_eps_bearer_settings_finish (_self, res, &error);
+    if (!properties) {
+        g_task_return_error (task, error);
+        g_object_unref (task);
+        return;
+    }
+
+    /* store result temporarily as task data */
+    g_task_set_task_data (task, properties, (GDestroyNotify) g_object_unref);
+
+    mm_base_modem_at_command (
+        MM_BASE_MODEM (self),
+        "+MGAUTH?",
+        20,
+        FALSE,
+        (GAsyncReadyCallback)load_initial_eps_bearer_settings_auth_ready,
+        task);
+}
+
+static void
+modem_3gpp_load_initial_eps_bearer_settings (MMIfaceModem3gpp    *self,
+                                             GAsyncReadyCallback  callback,
+                                             gpointer             user_data)
+{
+    GTask *task;
+
+    mm_obj_dbg_red (self, "modem_3gpp_load_initial_eps_bearer_settings");
+
+    task = g_task_new (self, NULL, callback, user_data);
+
+    /* First, use parent implementation to load the basic settings of the profile */
+    iface_modem_3gpp_parent->load_initial_eps_bearer_settings (
+        self,
+        (GAsyncReadyCallback)load_initial_eps_bearer_settings_parent_ready,
+        task);
+}
+
+/*****************************************************************************/
+/* Set initial EPS bearer settings */
+
+static gboolean
+modem_3gpp_set_initial_eps_bearer_settings_finish (MMIfaceModem3gpp  *self,
+                                                   GAsyncResult      *res,
+                                                   GError           **error)
+{
+    mm_obj_dbg_red (self, "modem_3gpp_set_initial_eps_bearer_settings_finish");
+    return g_task_propagate_boolean (G_TASK (res), error);
+}
+
+static void
+set_initial_eps_bearer_settings_parent_ready (MMIfaceModem3gpp *_self,
+                                              GAsyncResult     *res,
+                                              GTask            *task)
+{
+    MMBroadbandModemFibocomMc610 *self = MM_BROADBAND_MODEM_FIBOCOM_MC610 (_self);
+    GError                       *error = NULL;
+
+    mm_obj_dbg_red (self, "set_initial_eps_bearer_settings_parent_ready");
+
+    if (!mm_base_modem_at_command_finish (MM_BASE_MODEM(_self), res, &error)) {
+        mm_obj_warn (self, "couldn't configure initial EPS bearer auth settings: %s", error->message);
+        g_task_return_error (task, error);
+    } else
+        g_task_return_boolean (task, TRUE);
+    g_object_unref (task);
+}
+
+static void
+modem_3gpp_set_initial_eps_bearer_settings (MMIfaceModem3gpp    *_self,
+                                            MMBearerProperties  *properties,
+                                            GAsyncReadyCallback  callback,
+                                            gpointer             user_data)
+{
+    MMBroadbandModem *self = MM_BROADBAND_MODEM (_self);
+    GTask            *task;
+
+    task = g_task_new (self, NULL, callback, user_data);
+    g_task_set_task_data (task, g_object_ref (properties), (GDestroyNotify) g_object_unref);
+
+    mm_obj_dbg_red (_self, "modem_3gpp_set_initial_eps_bearer_settings, eps bearer cid: %d", mm_broadband_modem_get_initial_eps_bearer_cid(self));
+    mm_log_bearer_properties (_self, MM_LOG_LEVEL_DEBUG, "  ", properties);
+
+    /* First, use parent implementation to initialize the basic settings of the profile */
+    iface_modem_3gpp_parent->set_initial_eps_bearer_settings (
+        _self,
+        properties,
+        (GAsyncReadyCallback)set_initial_eps_bearer_settings_parent_ready,
+        task);
+}
+
+/*****************************************************************************/
+/* List profiles (3GPP profile management interface) */
+
+typedef struct {
+    GList *profiles;
+} ListProfilesContext;
+
+static void
+list_profiles_context_free (ListProfilesContext *ctx)
+{
+    mm_3gpp_profile_list_free (ctx->profiles);
+    g_slice_free (ListProfilesContext, ctx);
+}
+
+static gboolean
+modem_3gpp_profile_manager_list_profiles_finish (MMIfaceModem3gppProfileManager  *self,
+                                                 GAsyncResult                    *res,
+                                                 GList                          **out_profiles,
+                                                 GError                         **error)
+{
+    ListProfilesContext *ctx;
+
+    if (!g_task_propagate_boolean (G_TASK (res), error))
+        return FALSE;
+
+    ctx = g_task_get_task_data (G_TASK (res));
+    if (out_profiles)
+        *out_profiles = g_steal_pointer (&ctx->profiles);
+    return TRUE;
+}
+
+static void
+profile_manager_parent_list_profiles_ready (MMIfaceModem3gppProfileManager *self,
+                                            GAsyncResult                   *res,
+                                            GTask                          *task)
+{
+    ListProfilesContext          *ctx;
+    GError                       *error = NULL;
+    MMBroadbandModemFibocomMc610 *fibocom = MM_BROADBAND_MODEM_FIBOCOM_MC610 (self);
+    GList                        *l;
+    MM3gppProfile                *found = NULL;
+
+    mm_obj_dbg_red (self, "profile_manager_parent_list_profiles_ready");
+
+    ctx = g_slice_new0 (ListProfilesContext);
+    g_task_set_task_data (task, ctx, (GDestroyNotify) list_profiles_context_free);
+
+    if (!iface_modem_3gpp_profile_manager_parent->list_profiles_finish (self, res, &ctx->profiles, &error)) {
+        g_task_return_error (task, error);
+        g_object_unref (task);
+        return;
+    }
+
+    // look if profile0 is returned by the modem
+    if (ctx->profiles) {
+        for (l = ctx->profiles; l; l = l->next) {
+            MM3gppProfile *profile = MM_3GPP_PROFILE (l->data);
+            // TODO: replace 0 by mm_broadband_modem_get_initial_eps_bearer_cid (fibocom) ?
+            if (0 == mm_3gpp_profile_get_profile_id (profile)) {
+                found = g_object_ref (profile);
+                break;
+            }
+        }
+    }
+
+    if (found) {
+        // profile 0 found in modem, update local storage
+        mm_obj_dbg_red (self, "profile 0 fetched from modem");
+        if (fibocom->priv->profile0_stored && fibocom->priv->profile0) {
+            g_object_unref (fibocom->priv->profile0);
+        }
+        fibocom->priv->profile0 = g_object_ref (found);
+        fibocom->priv->profile0_stored = TRUE;
+    } else {
+        // profil 0 doesn't exist in modem, check if we have it locally
+        if (fibocom->priv->profile0_stored) {
+            mm_obj_dbg_red (self, "profile 0 fetched from local store");
+            ctx->profiles = g_list_prepend (ctx->profiles, g_object_ref (fibocom->priv->profile0));
+        } else {
+            // profile 0 not found in modem or local store, create a minimal profile 0
+            mm_obj_dbg_red (self, "profile 0 not found in modem or local store, creating minimal profile 0");
+            fibocom->priv->profile0 = mm_3gpp_profile_new ();
+            mm_3gpp_profile_set_profile_id (fibocom->priv->profile0, 0);
+            mm_3gpp_profile_set_apn (fibocom->priv->profile0, "internet");
+            fibocom->priv->profile0_stored = TRUE;
+            ctx->profiles = g_list_prepend (ctx->profiles, g_object_ref (fibocom->priv->profile0));
+        }
+    }
+    // if profile0 has not been stored already, we create a minimal profil
+    if (!fibocom->priv->profile0_stored) {
+        mm_obj_dbg_red (self, "creating minimal profile 0");
+        fibocom->priv->profile0 = mm_3gpp_profile_new ();
+        mm_3gpp_profile_set_profile_id (fibocom->priv->profile0, 0);
+        mm_3gpp_profile_set_apn (fibocom->priv->profile0, "internet");
+        fibocom->priv->profile0_stored = TRUE;
+    }
+
+    g_task_return_boolean (task, TRUE);
+    g_object_unref (task);
+}
+
+static void
+modem_3gpp_profile_manager_list_profiles (MMIfaceModem3gppProfileManager *self,
+                                          GAsyncReadyCallback             callback,
+                                          gpointer                        user_data)
+{
+    GTask *task;
+
+    mm_obj_dbg_red (self, "modem_3gpp_profile_manager_list_profiles");
+
+    task = g_task_new (self, NULL, callback, user_data);
+
+    iface_modem_3gpp_profile_manager_parent->list_profiles (
+        self,
+        (GAsyncReadyCallback)profile_manager_parent_list_profiles_ready,
+        task);
+}
+
+/*****************************************************************************/
+/* Check profiles (3GPP profile management interface) */
+
+typedef struct {
+    gboolean              new_id;
+    gint                  min_profile_id;
+    gint                  max_profile_id;
+    GEqualFunc            apn_cmp;
+    MM3gppProfileCmpFlags profile_cmp_flags;
+} CheckFormatContext;
+
+static void
+check_format_context_free (CheckFormatContext *ctx)
+{
+    g_slice_free (CheckFormatContext, ctx);
+}
+
+static gboolean
+modem_3gpp_profile_manager_check_format_finish (MMIfaceModem3gppProfileManager  *self,
+                                                GAsyncResult                    *res,
+                                                gboolean                        *new_id,
+                                                gint                            *min_profile_id,
+                                                gint                            *max_profile_id,
+                                                GEqualFunc                      *apn_cmp,
+                                                MM3gppProfileCmpFlags           *profile_cmp_flags,
+                                                GError                         **error)
+{
+    CheckFormatContext *ctx;
+
+    mm_obj_dbg_red (self, "modem_3gpp_profile_manager_check_format_finish");
+
+    if (!g_task_propagate_boolean (G_TASK (res), error))
+        return FALSE;
+
+    ctx = g_task_get_task_data (G_TASK (res));
+
+    mm_obj_dbg_red (self, "new_id: %d", ctx->new_id);
+
+    if (new_id)
+        *new_id = ctx->new_id;
+    if (min_profile_id)
+        *min_profile_id = (gint) ctx->min_profile_id;
+    if (max_profile_id)
+        *max_profile_id = (gint) ctx->max_profile_id;
+    if (apn_cmp)
+        *apn_cmp = ctx->apn_cmp;
+    if (profile_cmp_flags)
+        *profile_cmp_flags = ctx->profile_cmp_flags;
+    return TRUE;
+}
+
+static void
+profile_manager_parent_check_format_ready (MMIfaceModem3gppProfileManager *self,
+                                           GAsyncResult                   *res,
+                                           GTask                          *task)
+{
+    GError             *error = NULL;
+    CheckFormatContext *ctx;
+
+    mm_obj_dbg_red (self, "profile_manager_parent_check_format_ready");
+
+    ctx = g_task_get_task_data (task);
+
+    if (!iface_modem_3gpp_profile_manager_parent->check_format_finish (self,
+                                                                       res,
+                                                                       &ctx->new_id,
+                                                                       &ctx->min_profile_id,
+                                                                       &ctx->max_profile_id,
+                                                                       &ctx->apn_cmp,
+                                                                       &ctx->profile_cmp_flags,
+                                                                       &error)) {
+        g_task_return_error (task, error);
+    } else {
+        // keep parent implementation flags
+        //ctx->profile_cmp_flags &= ~MM_3GPP_PROFILE_CMP_FLAGS_NO_AUTH;
+        g_task_return_boolean (task, TRUE);
+    }
+    g_object_unref (task);
+}
+
+static void
+modem_3gpp_profile_manager_check_format (MMIfaceModem3gppProfileManager *self,
+                                         MMBearerIpFamily                ip_type,
+                                         GAsyncReadyCallback             callback,
+                                         gpointer                        user_data)
+{
+    GTask              *task;
+    CheckFormatContext *ctx;
+
+    mm_obj_dbg_red (self, "modem_3gpp_profile_manager_check_format");
+
+    task = g_task_new (self, NULL, callback, user_data);
+    ctx = g_slice_new0 (CheckFormatContext);
+    g_task_set_task_data (task, ctx, (GDestroyNotify)check_format_context_free);
+
+    iface_modem_3gpp_profile_manager_parent->check_format (
+        self,
+        ip_type,
+        (GAsyncReadyCallback)profile_manager_parent_check_format_ready,
+        task);
+}
+
+/*****************************************************************************/
+/* Set profile (3GPP profile management interface) */
+
+#define MGAUTH_SET_MAX_ATTEMPTS       3
+#define MGAUTH_SET_RETRY_TIMEOUT_SECS 1
+
+typedef struct {
+    MM3gppProfile       *profile;
+    gchar               *cmd;
+    gint                 profile_id;
+    guint                n_attempts;
+} StoreProfileContext;
+
+static void
+store_profile_context_free (StoreProfileContext *ctx)
+{
+    g_free (ctx->cmd);
+    g_clear_object (&ctx->profile);
+    g_slice_free (StoreProfileContext, ctx);
+}
+
+static gint
+modem_3gpp_profile_manager_store_profile_finish (MMIfaceModem3gppProfileManager  *self,
+                                                 GAsyncResult                    *res,
+                                                 gint                            *out_profile_id,
+                                                 MMBearerApnType                 *out_apn_type,
+                                                 GError                         **error)
+{
+    StoreProfileContext *ctx;
+
+    mm_obj_dbg_red (self, "modem_3gpp_profile_manager_store_profile_finish");
+
+    if (!g_task_propagate_boolean (G_TASK (res), error))
+        return FALSE;
+
+    ctx = g_task_get_task_data (G_TASK (res));
+    if (out_profile_id)
+        *out_profile_id = ctx->profile_id;
+    if (out_apn_type)
+        *out_apn_type = ctx->profile_id == 0 ? MM_BEARER_APN_TYPE_INITIAL : MM_BEARER_APN_TYPE_NONE;
+    return TRUE;
+}
+
+static void profile_manager_store_profile_auth_settings (GTask *task);
+
+static gboolean
+profile_manager_mgauth_set_retry (GTask *task)
+{
+    profile_manager_store_profile_auth_settings (task);
+    return G_SOURCE_REMOVE;
+}
+
+static void
+profile_manager_mgauth_set_ready (MMBaseModem  *self,
+                                   GAsyncResult *res,
+                                   GTask        *task)
+{
+    StoreProfileContext *ctx;
+    g_autoptr(GError)    error = NULL;
+
+
+    mm_obj_dbg_red (self, "profile_manager_mgauth_set_ready");
+
+    ctx = g_task_get_task_data (task);
+
+    if (!mm_base_modem_at_command_finish (self, res, &error)) {
+        /* Retry configuring the context. It sometimes fails with a 583
+         * error ["a profile (CID) is currently active"] if a connect
+         * is attempted too soon after a disconnect. */
+        if (ctx->n_attempts < MGAUTH_SET_MAX_ATTEMPTS) {
+            mm_obj_dbg (self, "couldn't store auth settings in profile '%d': %s; retrying...",
+                        ctx->profile_id, error->message);
+            g_timeout_add_seconds (MGAUTH_SET_RETRY_TIMEOUT_SECS, (GSourceFunc)profile_manager_mgauth_set_retry, task);
+            return;
+        }
+        g_task_return_error (task, g_steal_pointer (&error));
+    } else
+        g_task_return_boolean (task, TRUE);
+    g_object_unref (task);
+}
+
+static void
+profile_manager_store_profile_auth_settings (GTask *task)
+{
+    MMIfaceModem3gppProfileManager *self;
+    StoreProfileContext            *ctx;
+    g_autofree gchar               *cmd = NULL;
+
+    self = g_task_get_source_object (task);
+    ctx = g_task_get_task_data (task);
+
+    mm_obj_dbg_red (self, "profile_manager_store_profile_auth_settings");
+    mm_log_3gpp_profile (self, MM_LOG_LEVEL_DEBUG, "  ", ctx->profile);
+
+    if (!ctx->cmd) {
+        const gchar         *user;
+        const gchar         *password;
+        MMBearerAllowedAuth  allowed_auth;
+
+        user = mm_3gpp_profile_get_user (ctx->profile);
+        password = mm_3gpp_profile_get_password (ctx->profile);
+        allowed_auth = mm_3gpp_profile_get_allowed_auth (ctx->profile);
+
+        /* Both user and password are required; otherwise firmware returns an error */
+        if (!user || !password || allowed_auth == MM_BEARER_ALLOWED_AUTH_NONE) {
+            mm_obj_dbg (self, "not using authentication");
+            ctx->cmd = g_strdup_printf ("+MGAUTH=%d,0", ctx->profile_id);
+        } else {
+            g_autofree gchar *quoted_user = NULL;
+            g_autofree gchar *quoted_password = NULL;
+            guint             fibocom_auth;
+
+            if (allowed_auth == MM_BEARER_ALLOWED_AUTH_UNKNOWN) {
+                mm_obj_dbg (self, "using default (PAP) authentication method");
+                fibocom_auth = 1;
+            } else if (allowed_auth & MM_BEARER_ALLOWED_AUTH_PAP) {
+                mm_obj_dbg (self, "using PAP authentication method");
+                fibocom_auth = 1;
+            } else if (allowed_auth & MM_BEARER_ALLOWED_AUTH_CHAP) {
+                mm_obj_dbg (self, "using CHAP authentication method");
+                fibocom_auth = 2;
+            } else {
+                g_autofree gchar *str = NULL;
+
+                str = mm_bearer_allowed_auth_build_string_from_mask (allowed_auth);
+                g_task_return_new_error (task,
+                                         MM_CORE_ERROR,
+                                         MM_CORE_ERROR_UNSUPPORTED,
+                                         "Cannot use any of the specified authentication methods (%s)",
+                                         str);
+                g_object_unref (task);
+                return;
+            }
+
+            quoted_user     = mm_at_quote_string (user);
+            quoted_password = mm_at_quote_string (password);
+            ctx->cmd = g_strdup_printf ("+MGAUTH=%d,%u,%s,%s",
+                                        ctx->profile_id,
+                                        fibocom_auth,
+                                        quoted_user,
+                                        quoted_password);
+        }
+    }
+
+    ctx->n_attempts++;
+    mm_base_modem_at_command (MM_BASE_MODEM (self),
+                              ctx->cmd,
+                              6,
+                              FALSE,
+                              (GAsyncReadyCallback)profile_manager_mgauth_set_ready,
+                              task);
+}
+
+static void
+profile_manager_parent_store_profile_ready (MMIfaceModem3gppProfileManager *self,
+                                            GAsyncResult                   *res,
+                                            GTask                          *task)
+{
+    GError *error = NULL;
+
+    mm_obj_dbg_red (self, "profile_manager_parent_store_profile_ready");
+
+    if (!iface_modem_3gpp_profile_manager_parent->store_profile_finish (self, res, NULL, NULL, &error)) {
+        g_task_return_error (task, error);
+        g_object_unref (task);
+        return;
+    }
+
+    profile_manager_store_profile_auth_settings (task);
+}
+
+static void
+modem_3gpp_profile_manager_store_profile (MMIfaceModem3gppProfileManager  *self,
+                                          MM3gppProfile                   *profile,
+                                          const gchar                     *index_field,
+                                          GAsyncReadyCallback              callback,
+                                          gpointer                         user_data)
+{
+    StoreProfileContext *ctx;
+    GTask               *task;
+    guint                profile_id;
+
+
+
+    MMBroadbandModemFibocomMc610 *self_mm = MM_BROADBAND_MODEM_FIBOCOM_MC610 (self);
+
+    g_assert (g_strcmp0 (index_field, "profile-id") == 0);
+
+    profile_id = mm_3gpp_profile_get_profile_id (profile);
+    mm_obj_dbg_red (self, "modem_3gpp_profile_manager_store_profile, profile id: %d", profile_id);
+    mm_log_3gpp_profile (self, MM_LOG_LEVEL_DEBUG, "  ", profile);
+
+    if (profile_id == 0) {
+        mm_obj_dbg_red (self, "Storing profile 0 for later use");
+
+        if (self_mm->priv->profile0) {
+            g_object_unref (self_mm->priv->profile0);
+        }
+
+        self_mm->priv->profile0 = g_object_ref (profile);
+        self_mm->priv->profile0_stored = TRUE;
+    }
+
+    task = g_task_new (self, NULL, callback, user_data);
+    ctx = g_slice_new0 (StoreProfileContext);
+    ctx->profile = g_object_ref (profile);
+    ctx->profile_id = profile_id;
+    g_assert (ctx->profile_id != MM_3GPP_PROFILE_ID_UNKNOWN);
+    g_task_set_task_data (task, ctx, (GDestroyNotify) store_profile_context_free);
+
+    iface_modem_3gpp_profile_manager_parent->store_profile (
+        self,
+        profile,
+        index_field,
+        (GAsyncReadyCallback)profile_manager_parent_store_profile_ready,
+        task);
+}
+
+/*****************************************************************************/
+/* Common operation to load expected CID for the initial EPS bearer */
+
+static gint
+load_initial_eps_bearer_cid_finish (MMBroadbandModem  *self,
+                                    GAsyncResult      *res,
+                                    GError           **error)
+{
+    return g_task_propagate_int (G_TASK (res), error);
+}
+
+static void
+load_initial_eps_bearer_cid (MMBroadbandModem    *self,
+                             GAsyncReadyCallback  callback,
+                             gpointer             user_data)
+{
+    GTask *task;
+
+    mm_obj_dbg_red (self, "Fibocom MC610: Using fixed CID 0 for initial EPS bearer");
+
+    task = g_task_new (self, NULL, callback, user_data);
+
+    /* Always return CID 0 for the initial EPS bearer */
+    g_task_return_int (task, 0);
+    g_object_unref (task);
+}
+
+
 /*****************************************************************************/
 
 static const gchar *primary_init_sequence[] = {
@@ -713,10 +1410,33 @@ iface_modem_firmware_init (MMIfaceModemFirmwareInterface *iface)
     iface->load_update_settings_finish = mm_shared_fibocom_firmware_load_update_settings_finish;
 }
 
-static MMIfaceModemFirmwareInterface *
-peek_parent_firmware_interface (MMSharedFibocom *self)
+static void
+iface_modem_3gpp_init (MMIfaceModem3gppInterface *iface)
 {
-    return iface_modem_firmware_parent;
+    iface_modem_3gpp_parent = g_type_interface_peek_parent (iface);
+
+    iface->load_initial_eps_bearer_settings = modem_3gpp_load_initial_eps_bearer_settings;
+    iface->load_initial_eps_bearer_settings_finish = modem_3gpp_load_initial_eps_bearer_settings_finish;
+    iface->set_initial_eps_bearer_settings = modem_3gpp_set_initial_eps_bearer_settings;
+    iface->set_initial_eps_bearer_settings_finish = modem_3gpp_set_initial_eps_bearer_settings_finish;
+}
+
+static void
+iface_modem_3gpp_profile_manager_init (MMIfaceModem3gppProfileManagerInterface *iface)
+{
+    iface_modem_3gpp_profile_manager_parent = g_type_interface_peek_parent (iface);
+
+    iface->list_profiles = modem_3gpp_profile_manager_list_profiles;
+    iface->list_profiles_finish = modem_3gpp_profile_manager_list_profiles_finish;
+    iface->check_format = modem_3gpp_profile_manager_check_format;
+    iface->check_format_finish = modem_3gpp_profile_manager_check_format_finish;
+    /* note: the parent check_activated_profile() implementation using +CGACT? seems to
+     * be perfectly valid. */
+    //iface->deactivate_profile = modem_3gpp_profile_manager_deactivate_profile;
+    //iface->deactivate_profile_finish = modem_3gpp_profile_manager_deactivate_profile_finish;
+    //iface->check_activated_profile
+    iface->store_profile = modem_3gpp_profile_manager_store_profile;
+    iface->store_profile_finish = modem_3gpp_profile_manager_store_profile_finish;
 }
 
 static MMBaseModemClass *
@@ -725,6 +1445,12 @@ peek_parent_class (MMSharedFibocom *self)
     return MM_BASE_MODEM_CLASS (mm_broadband_modem_fibocom_mc610_parent_class);
 }
 
+static MMIfaceModemFirmwareInterface *
+peek_parent_firmware_interface (MMSharedFibocom *self)
+{
+    return iface_modem_firmware_parent;
+}
+
 static void
 shared_fibocom_init (MMSharedFibocomInterface *iface)
 {
@@ -743,4 +1469,6 @@ mm_broadband_modem_fibocom_mc610_class_init (MMBroadbandModemFibocomMc610Class *
     /* Virtual methods */
     object_class->finalize = finalize;
     broadband_modem_class->setup_ports = setup_ports;
+    broadband_modem_class->load_initial_eps_bearer_cid = load_initial_eps_bearer_cid;
+    broadband_modem_class->load_initial_eps_bearer_cid_finish = load_initial_eps_bearer_cid_finish;
 }
-- 
2.45.2

