From 6675c77ee7fa9bab86cc8fc2ca8a58fb106ea322 Mon Sep 17 00:00:00 2001
From: Anton Beitler <abeitler@semtech.com>
Date: Thu, 6 Jan 2022 16:38:01 +0100
Subject: [PATCH] Update mbedTLS dependency to version 2.7.5. Fixes #142

---
 deps/mbedtls/prep.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/deps/mbedtls/prep.sh b/deps/mbedtls/prep.sh
index 2875cd7..4f774f0 100755
--- a/deps/mbedtls/prep.sh
+++ b/deps/mbedtls/prep.sh
@@ -30,7 +30,7 @@ set -e
 cd $(dirname $0)
 
 if [[ ! -d git-repo ]]; then
-    git clone -b mbedtls-2.6 --single-branch --depth 1 https://github.com/ARMmbed/mbedtls.git git-repo
+    git clone -b mbedtls-2.7.5 --single-branch --depth 1 https://github.com/ARMmbed/mbedtls.git git-repo
 fi
 
 if [[ -z "$platform" ]] || [[ -z "$variant" ]]; then
-- 
2.32.0