From 7139bf507e4b4a3fe6a5db6407e4693521c3db7a Mon Sep 17 00:00:00 2001
From: Yannick Lanz <yannick.lanz@wifx.net>
Date: Wed, 26 Mar 2025 16:01:59 +0100
Subject: [PATCH 5/6] fibocom: cleanup load unlock retries callback

Signed-off-by: Yannick Lanz <yannick.lanz@wifx.net>
---
 .../fibocom/mm-broadband-modem-fibocom.c      | 58 ++++++++-----------
 1 file changed, 24 insertions(+), 34 deletions(-)

diff --git a/src/plugins/fibocom/mm-broadband-modem-fibocom.c b/src/plugins/fibocom/mm-broadband-modem-fibocom.c
index d9fc7ead..08a9fdc8 100644
--- a/src/plugins/fibocom/mm-broadband-modem-fibocom.c
+++ b/src/plugins/fibocom/mm-broadband-modem-fibocom.c
@@ -365,51 +365,41 @@ set_current_modes (MMIfaceModem        *_self,
 /* Load unlock retries (Modem interface) */
 
 static MMUnlockRetries *
-load_unlock_retries_finish (MMIfaceModem *self,
-                            GAsyncResult *res,
-                            GError **error)
+modem_load_unlock_retries_finish (MMIfaceModem *self,
+                                  GAsyncResult *res,
+                                  GError **error)
 {
-    MMUnlockRetries *unlock_retries;
     const gchar *response;
-    gint matched;
-    guint a, b, c ,d;
+    guint pin1, puk1, pin2 ,puk2;
 
     response = mm_base_modem_at_command_finish (MM_BASE_MODEM (self), res, error);
     if (!response)
         return NULL;
 
-    matched = sscanf (response, "+TPIN: %d,%d,%d,%d",
-                      &a, &b, &c, &d);
-    if (matched != 4) {
-        g_set_error (error,
-                     MM_CORE_ERROR,
-                     MM_CORE_ERROR_FAILED,
-                     "Could not parse PIN retries results: '%s'",
-                     response);
-        return NULL;
-    }
+    response = mm_strip_tag (response, "+TPIN:");
+    if (sscanf (response, " %d,%d,%d,%d", &pin1, &puk1, &pin2, &puk2) == 4) {
+        MMUnlockRetries *unlock_retries;
+        unlock_retries = mm_unlock_retries_new ();
+        mm_unlock_retries_set (unlock_retries, MM_MODEM_LOCK_SIM_PIN, pin1);
+        mm_unlock_retries_set (unlock_retries, MM_MODEM_LOCK_SIM_PUK, puk1);
+        mm_unlock_retries_set (unlock_retries, MM_MODEM_LOCK_SIM_PIN2, pin2);
+        mm_unlock_retries_set (unlock_retries, MM_MODEM_LOCK_SIM_PUK2, puk2);
 
-    if (a > 998) {
-        g_set_error (error,
-                     MM_CORE_ERROR,
-                     MM_CORE_ERROR_FAILED,
-                     "Invalid PIN attempts left: '%u'",
-                     a);
-        return NULL;
+        return unlock_retries;
     }
 
-    unlock_retries = mm_unlock_retries_new ();
-    mm_unlock_retries_set (unlock_retries, MM_MODEM_LOCK_SIM_PIN, a);
-    mm_unlock_retries_set (unlock_retries, MM_MODEM_LOCK_SIM_PUK, b);
-    mm_unlock_retries_set (unlock_retries, MM_MODEM_LOCK_SIM_PIN2, c);
-    mm_unlock_retries_set (unlock_retries, MM_MODEM_LOCK_SIM_PUK2, d);
-    return unlock_retries;
+    g_set_error (error,
+        MM_CORE_ERROR,
+        MM_CORE_ERROR_FAILED,
+        "Could not parse PIN retries results: '%s'",
+        response);
+    return NULL;
 }
 
 static void
-load_unlock_retries (MMIfaceModem *self,
-                     GAsyncReadyCallback callback,
-                     gpointer user_data)
+modem_load_unlock_retries (MMIfaceModem *self,
+                           GAsyncReadyCallback callback,
+                           gpointer user_data)
 {
     mm_base_modem_at_command (MM_BASE_MODEM (self),
                               "+TPIN?",
@@ -1037,8 +1027,8 @@ iface_modem_init (MMIfaceModem *iface)
     iface->load_current_modes_finish = load_current_modes_finish;
     iface->set_current_modes = set_current_modes;
     iface->set_current_modes_finish = set_current_modes_finish;
-    iface->load_unlock_retries = load_unlock_retries;
-    iface->load_unlock_retries_finish = load_unlock_retries_finish;
+    iface->load_unlock_retries = modem_load_unlock_retries;
+    iface->load_unlock_retries_finish = modem_load_unlock_retries_finish;
     iface->setup_sim_hot_swap = modem_setup_sim_hot_swap;
     iface->setup_sim_hot_swap_finish = modem_setup_sim_hot_swap_finish;
     iface->cleanup_sim_hot_swap = modem_cleanup_sim_hot_swap;
-- 
2.34.1

