From 9b0a81bcbd103e396dac34633e1ef267d8e75a52 Mon Sep 17 00:00:00 2001
From: Yannick Lanz <yannick.lanz@wifx.net>
Date: Fri, 13 Aug 2021 10:14:56 +0200
Subject: [PATCH 6/7] l1: disable unused PWM and tcb timer support

Signed-off-by: Yannick Lanz <yannick.lanz@wifx.net>
---
 arch/arm/boot/dts/l1.dts | 17 ++++++++---------
 1 file changed, 8 insertions(+), 9 deletions(-)

diff --git a/arch/arm/boot/dts/l1.dts b/arch/arm/boot/dts/l1.dts
index fe0852a0c392..462077219d74 100644
--- a/arch/arm/boot/dts/l1.dts
+++ b/arch/arm/boot/dts/l1.dts
@@ -250,16 +250,15 @@
 				status = "okay";
 			};
 
+			/* disable unused TCB */
+			tcb0: timer@f801c000 {
+				status = "disabled";
+			};
+			tcb1: timer@fc020000 {
+				status = "disabled";
+			};
 			tcb2: timer@fc024000 {
-				timer0: timer@0 {
-					compatible = "atmel,tcb-timer";
-					reg = <0>;
-				};
-
-				timer1: timer@1 {
-					compatible = "atmel,tcb-timer";
-					reg = <1>;
-				};
+				status = "disabled";
 			};
 
 			watchdog@fc068640 {
-- 
2.25.1

