From 86a150c0751ba2ea3f3977ece5c31a427886d785 Mon Sep 17 00:00:00 2001
From: OpenEmbedded <oe.patch@oe>
Date: Mon, 18 Oct 2021 21:26:37 +0200
Subject: [PATCH 10/11] l1.dts: rename serial LED to service

Signed-off-by: Yannick Lanz <yannick.lanz@wifx.net>
---
 arch/arm/boot/dts/l1.dts | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/arch/arm/boot/dts/l1.dts b/arch/arm/boot/dts/l1.dts
index 006b2fd851f7..2aae8d140261 100644
--- a/arch/arm/boot/dts/l1.dts
+++ b/arch/arm/boot/dts/l1.dts
@@ -155,15 +155,15 @@
 						#size-cells = <0>;
 						status = "okay";
 
-						sysled {
+						statusled {
 							reg = <0>;
 							label = "status";
 							linux,default-trigger = "heartbeat";
 						};
 
-						serled {
+						serviceled {
 							reg = <1>;
-							label = "serial";
+							label = "service";
 							linux,default-trigger = "wgw-usbc-data-mode";
 							trigger-sources = <&usbc_ctrl>;
 						};
-- 
2.25.1

